src/midst_toolkit/attacks/tf/data_utils.py

```
@@ -0,0 +1,304 @@
+    missing_keys_challenge = [key for key in keys_for_deduplication if key not in df_challenge.columns]
+    if missing_keys_merge or missing_keys_challenge:
+        raise ValueError(f"Missing columns for deduplication: {missing_keys_merge + missing_keys_challenge}")
+
+    df_merge_without_challenge = df_merge[
+        ~df_merge.set_index(keys_for_deduplication).index.isin(df_challenge.set_index(keys_for_deduplication).index)
+    ]
+
+    return df_merge_without_challenge, df_challenge, df_challenge_labels
+
+
+def get_tpr_at_fpr(true_membership: list[int], predictions: list[float], max_fpr: float = 0.1) -> float:
+    """
+    Calculates the best True Positive Rate when the False Positive Rate is at most `max_fpr`.
+    """
+    fpr, tpr, _ = roc_curve(true_membership, predictions)
+    valid_tpr = tpr[fpr <= max_fpr]
+    if len(valid_tpr) == 0:
+        return 0.0
+    return float(max(valid_tpr))
+
+
+def evaluate_attack_performance(
+    indices: list[int] | None,
+    description: str,
+    tabddpm_data_dir: Path,
+    model_type: str,
+    predictions_file_name: str,
+) -> dict[str, float]:
+    if indices is None or len(indices) == 0:
+        raise ValueError(f"Indices list is empty for {description}. Cannot evaluate attack performance.")
+
+    predictions: list[np.ndarray] = []
+    solutions: list[np.ndarray] = []
+    for model_number in tqdm(indices, desc=f"Evaluating on {description} models", unit="model"):
+        model_folder = f"{model_type}_{model_number}"
+        path = tabddpm_data_dir / model_folder
+        try:
+            predictions.append(np.loadtxt(path / predictions_file_name))
+            solutions.append(np.loadtxt(path / "challenge_label.csv", skiprows=1))
+        except FileNotFoundError:
+            print(f"Warning: Prediction or label file not found for model {model_number}. Skipping.")
+            continue
+
+    if not predictions:
+        print(f"No predictions found for {description} models.")
```

Switch to logging here and I would throw an error if this is the case rather than returning something?